### PR TITLE
[Flux] Enable checkpointing

### PR DESCRIPTION
**Pull request by wwwjn**
## Context:
1. Change flux-dev / flux-schnell model training to be ~30000 step based on current MAST training results
2. Enable checkpointing. We enabled final_layer reshard_after_forward to solve issue described [here](https://github.com/pytorch/torchtitan/pull/1167#discussion_r2075967381)

## Test
If we run following 2 runs, the training loss curve should be identical with `deterministic = True`:
1. Without checkpoint save and load, total step=10
2. Save checkpoint at step 5, and load checkpoint at step 5, continue training

Currently issue #1194 makes the training loss not strictly identical. To exclude the influence of #1194, we reset the seeds (by calling `set_deterministic()` at the beginning of step 6. Then the checkpoint save/load makes the training loss identical.

Screenshot 2025-05-14 at 2 06 23 PM

### CODE REVIEW

File: torchtitan/experiments/flux/README.md
Comment (tianyu-l) [inline torchtitan/experiments/flux/README.md:1]: Let's remove checkpointing from the TODO item. And also add a paragraph on right now, why we don't expect resuming from checkpoint yielding the same loss curves, due to randomness not being checkpointed. I think we should fix this.

File: torchtitan/experiments/flux/train.py [hunk 1]
```diff
@@ -182,7 +182,11 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
             or self.step == self.job_config.training.steps
         ):
             model.eval()
+            # We need to set reshard_after_forward before last forward pass.
+            # So the model wieghts are sharded the same way for checkpoint saving.
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py:186]: Hmm I still didn't fully understand. What if we don't do this -- what failure would we hit?

How is it related to checkpointing, as you are setting it around eval.

Comment (fegin) [reply torchtitan/experiments/flux/train.py:186]: This is because `checkpointing` is run after eval. Without the settings, FSDP will not reshard meaning the weights will be full weights, which cause checkpointing failures.

**Review by fegin — COMMENTED** [review]
I suggest that you split the "fix" of parallelize_flux.py into another PR. And I thought you plan to checkpoint the rng state?

File: torchtitan/experiments/flux/train.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/train.py:186]: oh, now I see. The reason is that eval step only runs forward, but not backward, so FSDP `reshard_after_forward` plan would interfere with how parameters look like for the potential subsequent checkpointing step.

It is not good to couple FSDP wrapping's `reshard_after_forward` details with `eval_step`, and `train.py` in general. We should think of a cleaner way to do eval.

For now let's put a TODO item.

**Review by tianyu-l — APPROVED** [review]
LGTM. Please address comments before merge.

**Comment by wwwjn** [comment]
> I suggest that you split the "fix" of parallelize_flux.py into another PR.

Thank you @fegin for reminding, I created a separate PR for this. I will make separate PRs for changes for later changes as well, thanks for reminding!
